### ISSUE TITLE
[REVIEW]: AdaptiveResonance.jl: A Julia implementation of Adaptive Resonance Theory (ART) algorithms

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3671]
**Submitting author:** @AP6YC (<a href="http://orcid.org/0000-0003-2442-8901">Sasha Petrenko</a>)
**Repository:** <a href="https://github.com/AP6YC/AdaptiveResonance.jl" target ="_blank">https://github.com/AP6YC/AdaptiveResonance.jl</a>
**Version:** 0.2.2
**Editor:** @adi3
**Reviewer:** @aaronpeikert, @hayesall
**Archive:** Pending

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/4db5d132c25546b235f167cbbfdd4af5/status.svg)](https://joss.theoj.org/papers/4db5d132c25546b235f167cbbfdd4af5)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/4db5d132c25546b235f167cbbfdd4af5"><img src="https://joss.theoj.org/papers/4db5d132c25546b235f167cbbfdd4af5/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/4db5d132c25546b235f167cbbfdd4af5/status.svg)](https://joss.theoj.org/papers/4db5d132c25546b235f167cbbfdd4af5)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/AP6YC/AdaptiveResonance.jl" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@aaronpeikert & @hayesall, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @adi3 know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @aaronpeikert
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/AP6YC/AdaptiveResonance.jl">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/AP6YC/AdaptiveResonance.jl">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@AP6YC) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @hayesall
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/AP6YC/AdaptiveResonance.jl">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/AP6YC/AdaptiveResonance.jl">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@AP6YC) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @aaronpeikert, @hayesall it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1051/0004-6361/201629272 is OK
- 10.1051/0004-6361/201322068 is OK
- 10.1016/j.neunet.2012.09.017 is OK
- 10.1037/0033-295X.87.1.1 is OK
- 10.1016/j.neunet.2019.09.012 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.03 s (1186.3 files/s, 130118.1 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Julia                           25            470            503           2366
Markdown                         7            193              0            510
YAML                             6             23            132            164
TeX                              1              7              0             97
TOML                             2              4              0             28
-------------------------------------------------------------------------------
SUM:                            41            697            635           3165
-------------------------------------------------------------------------------


Statistical information for the repository '72e2445e09bf5034e5ac9445' was
gathered on 2021/08/30.
No commited files with the specified extensions were found.
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03671/joss.03671/10.21105.joss.03671.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03671/joss.03671/10.21105.joss.03671.pdf) :page_facing_up: :point_left:

**Comment by adi3** [comment]
👋🏼 @AP6YC this is the review thread for the paper. All of our communications will happen here from now on.

👋🏼  @hayesall @aaronpeikert - you both have checklists at the top of this thread with the JOSS requirements. As you go over the submission, please check any items that you feel have been satisfied. There are also links to the JOSS reviewer guidelines.

Here is a little more context for first-time reviewers :) - The JOSS review is different from most other journals. Our goal is to work with the authors to help them meet our criteria instead of merely passing judgment on the submission. As such, the reviewers are encouraged to submit issues and pull requests on the software repository. When doing so, please mention #3671 so that a link is created to this thread (and I can keep an eye on what is happening). Please also feel free to comment and ask questions on this thread. In my experience, it is better to post comments/questions/suggestions as you come across them instead of waiting until you've reviewed the entire package.

We aim for reviews to be completed within about 2-4 weeks. Please let me know if any of you require some more time. We can also use Whedon (our bot) to set automatic reminders if you know you'll be away for a known period of time.

Please feel free to ping me (@adi3) if you have any questions/concerns. Thank you for all your help!!

**Comment by whedon** [comment]
:wave: @aaronpeikert, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @hayesall, please update us on how your review is going (this is an automated reminder).

**Comment by hayesall** [comment]
Hey @adi3! I saw your note a week ago, but I haven't started yet (beyond reading some of the docs). I'm planning to do my first pass this Saturday.

**Comment by aaronpeikert** [comment]
I was on vacation till last week and are now on the receiving end of a review but this is the next thing on my to-do list.

**Comment by adi3** [comment]
@aaronpeikert no worries, enjoy your vacation! Speak to you when you're back

**Comment by adi3** [comment]
@AP6YC please address the issues listed above that require your attention. Can you get these done within this month? I'd like to take this review forward.

**Comment by AP6YC** [comment]
@adi3 Sure thing, end of the month is very feasible. I have been tackling the issues on the project's issues page, but I will be more diligent to give updates on them here too.

**Comment by adi3** [comment]
@AP6YC - please update us on your progress, and an estimated date for completion. Thank you!

**Comment by AP6YC** [comment]
You've got it; for starters, [PR 46](https://github.com/AP6YC/AdaptiveResonance.jl/pull/46) answers [Issue 24](https://github.com/AP6YC/AdaptiveResonance.jl/issues/24): the target audience within the statement of need.

These and other changes are in `develop` since switching to GitFlow, and I am hoping to make a major release once I have addressed all of the reviewers' concerns and recommendations.

**Comment by AP6YC** [comment]
[PR 48](https://github.com/AP6YC/AdaptiveResonance.jl/pull/48/) addresses [Issue 41](https://github.com/AP6YC/AdaptiveResonance.jl/issues/41): method redefinition (a bug, but listed here for completeness).

**Comment by AP6YC** [comment]
[PR 39](https://github.com/AP6YC/AdaptiveResonance.jl/pull/39) partially addresses [Issue 26](https://github.com/AP6YC/AdaptiveResonance.jl/issues/26): the "Hello World" example. An additional PR will come to clean up the `examples/` directory into the docs, which will also address [Issue 27](https://github.com/AP6YC/AdaptiveResonance.jl/issues/27): basic working examples.

**Comment by AP6YC** [comment]
As promised, PR #49 now addresses Issues [26](https://github.com/AP6YC/AdaptiveResonance.jl/issues/26) [27](https://github.com/AP6YC/AdaptiveResonance.jl/issues/27) by implementing many working examples using the `DemoCards` package in the Examples section of the docs and removing the deprecated example page and folder.

**Comment by adi3** [comment]
@hayesall @aaronpeikert - can you please continue with your reviews in light of the latest fixes done by Sasha? Thank you!

**Comment by aaronpeikert** [comment]
I'll have a look.

**Comment by hayesall** [comment]
Yes! I've checked off items as they've come in.

I'll review functionality again (there was some discussion on assumptions in https://github.com/AP6YC/AdaptiveResonance.jl/issues/42 that have now been clarified).

**Comment by aaronpeikert** [comment]
@AP6YC  I am almost satisfied (as soon as you merge dev and formally release a new version). The only thing that prevents me from ticking a box is https://github.com/AP6YC/AdaptiveResonance.jl/issues/23

While I do not consider them blocking, https://github.com/AP6YC/AdaptiveResonance.jl/issues/28 & https://github.com/AP6YC/AdaptiveResonance.jl/issues/25 would really increase the value of the paper/software.

**Comment by adi3** [comment]
@AP6YC can you address the final few issues brought up by @aaronpeikert ? I also see some of the issues opened by @hayesall as still open so please tend to them as well once you have a chance. We're on the home stretch here and need just a tiny more effort to get this across the finish line!

**Comment by adi3** [comment]
@aaronpeikert @hayesall - looks like your review is only left with the *Software paper* portion. Is everything in the code repository to your satisfaction at this point? Would be great if we can have the reviews finished sometime this month. Can you please shoot for it? Thanks in advance!

**Comment by adi3** [comment]
@editorialbot remind @aaronpeikert @hayesall  in 2 weeks

**Comment by aaronpeikert** [comment]
@AP6YC Could you give us an update on this? Can we somehow support you in finishing the software paper?

As @adi3 says:

> We're on the home stretch here and need just a tiny more effort to get this across the finish line!

**Comment by adi3** [comment]
@AP6YC - we have not heard from you in close to 4 months. Without further participation in this thread, I'm afraid we'll have to reject this submission after all this effort from your (and the reviewers!) end. Please contact us.

**Comment by AP6YC** [comment]
All,

I sincerely apologize for my absence in this project and this thread, which despite life events I cannot excuse. I would like to thank @adi3 for reaching out to me via email to correct my neglect of this thread. To quickly conclude this review, and as a show of good faith to everyone's time, I have gone through this thread and listed out the remaining reviewers' concerns:

- [ ] https://github.com/AP6YC/AdaptiveResonance.jl/issues/23
- [ ] https://github.com/AP6YC/AdaptiveResonance.jl/issues/28
- [ ] https://github.com/AP6YC/AdaptiveResonance.jl/issues/25

In summary, I will make immediate changes to the paper itself to compare and contrast existing implementations of ART algorithms, and I will complete the work that I began before in formalizing the docstrings that document the various options for each ART module in the index.

My next message will be to show that these last concerns are completed for the reviewers.

**Comment by hayesall** [comment]
@whedon generate pdf

**Comment by hayesall** [comment]
@editorialbot generate pdf

**Comment by hayesall** [comment]
Just about there! I've ticked off all my boxes.

A few things need fixed in the references, these may not be exhaustive:

---

![image](https://user-images.githubusercontent.com/11916674/164799656-5e030942-c010-4c96-be54-f081ac43a218.png)

- Adjust `author`

---

![image](https://user-images.githubusercontent.com/11916674/164799849-44716e3f-c2eb-4706-b25e-5a93cb0157c5.png)

- Issue number `(xxxx)` probably isn't correct.

---

![image](https://user-images.githubusercontent.com/11916674/164799552-0f8b878e-60ec-437a-abed-155f904deca0.png)

- Remove the space between `Code ?` in `title`

---

![image](https://user-images.githubusercontent.com/11916674/164798709-63b4e38d-73f4-4eb2-8099-4712aff1f953.png)

- Drop "Scholar's Mine" from `title`
- Add the DOI: https://doi.org/10.13140/RG.2.2.11823.25761

---

![image](https://user-images.githubusercontent.com/11916674/164799272-fd09ada9-14a1-4334-9acc-6c9d928f2c9d.png)

- `author` section needs adjusted 

---

![image](https://user-images.githubusercontent.com/11916674/164799065-e6e3a7a0-a31e-42e9-a17a-57a2cea107c8.png)

- `(xxxx)` seems unlikely to be the correct issue number

**Comment by AP6YC** [comment]
@hayesall All very good catches! Sorry for the inconsistencies in the bibliography. I addressed your concerns and more, such as software citations, with a recent hotfix into `master`.

@editorialbot generate pdf

**Comment by AP6YC** [comment]
@editorialbot generate pdf

**Comment by adi3** [comment]
@editorialbot check references

**Comment by adi3** [comment]
@AP6YC great work! Looks like everything is set.  I need one final formality to be done from your side. Please make a tagged release of the project and archive it. Then report the version number and archive DOI here. Once I have that, I will send this for publication.

@hayesall @aaronpeikert thank you for all your help and effort with this review!

**Comment by AP6YC** [comment]
@adi3 I have archived the project at Zenodo. The release folding in all of the last changes has the following record, DOI, and version:
- Record: https://zenodo.org/record/6506781
- DOI: 10.5281/zenodo.6506781
- Version: v0.5.0

Does this work, or should I archive in another way?

**Comment by adi3** [comment]
@editorialbot set 10.5281/zenodo.6506781 as archive

**Comment by adi3** [comment]
@editorialbot set v0.5.0 as version

**Comment by adi3** [comment]
@editorialbot recommend-accept

**Comment by kyleniemeyer** [comment]
Hi @AP6YC, I am the AEIC on duty this week doing some final checks before publishing.

I noticed that the author list of your Zenodo archive does not match the author list of the paper. Generally, these should match. Can you either clean up the author list in the Zenodo metadata or justify why these should be different?

**Comment by AP6YC** [comment]
@kyleniemeyer Thank you for catching that! I suspect that the discrepancy exists because I folded in two PRs done by the two editors of the paper before the latest release, which pulled their names into the automatic Zenodo upload, and I overlooked adding another author to Zenodo in the first place.

I will clean up the upload metadata to correct the authors to reflect the paper. Will it be appropriate to move the two editors as contributors to the Zenodo upload, or should they be removed entirely?

**Comment by kyleniemeyer** [comment]
@AP6YC if they did not contribute to the software itself, then I think it's most appropriate to just remove them entirely, but it's up to you

**Comment by AP6YC** [comment]
@kyleniemeyer gotchya; I added them as "other" contributors since they did make software contributions. The author list for the upload is fixed now. Thanks again for pointing it out!

**Comment by kyleniemeyer** [comment]
@editorialbot accept

**Comment by kyleniemeyer** [comment]
Congratulations @AP6YC on your article's publication in JOSS!

Many thanks to @aaronpeikert and @hayesall for reviewing this, and @adi3 for editing.
